src/Avalonia.Base/PropertyStore/ValueStore.cs

```diff
@@ -283,11 +284,25 @@ public void SetLocalValue<T>(StyledProperty<T> property, T value)
 
         public T GetValue<T>(StyledProperty<T> property)
         {
+            // <!> Performance critical method
             if (_effectiveValues.TryGetValue(property, out var v))
-                return ((EffectiveValue<T>)v).Value;
+                return CastEffectiveValue<T>(v).Value;
             if (property.Inherits && TryGetInheritedValue(property, out v))
-                return ((EffectiveValue<T>)v).Value;
-            return property.GetDefaultValue(Owner.GetType());
+                return CastEffectiveValue<T>(v).Value;
+            return property.GetDefaultValue(Owner);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static EffectiveValue<T> CastEffectiveValue<T>(EffectiveValue value)
+        {
+#if DEBUG
+            return (EffectiveValue<T>)value;
+#else
+            // Avoid casts in release mode for performance since GetValue is a very hot path.
+            // We control every path:
+            // it shouldn't be possible to have something else than an EffectiveValue<T> stored for a T property.
+            return Unsafe.As<EffectiveValue<T>>(value);
+#endif
```

Comment:
Is there any debugging disadvantage to keep the same code in DEBUG configuration?

Reply:
Yes, safety :) The normal cast acts as an assertion here: I think it's better to catch potential logic errors earlier in debug than reading/writing arbitrary memory. Performance won't matter in debug since the code is already un-optimized.

Note that I've been very careful to use `Unsafe.As` only because the type should be 100% controlled and we're guaranteed to be single threaded (via the `CheckAccess()` call). Even with that, I like having the cast as safety net while making code changes.